packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -0,0 +1,457 @@
+   * Returning `false` here prevents closing modal.
+   */
+  onClose?(event: MouseEvent): void | boolean;
+
+  /**
+   * Called for all `onKeyDown` events that do not close the modal
+   */
+  onKeyDown(evt: KeyboardEvent): void;
+
+  /**
+   * Specify whether the Modal is currently open
+   */
+  open?: boolean;
+
+  preventCloseOnClickOutside?: boolean;
+
+  /**
+   * Specify a CSS selector that matches the DOM element that should be
+   * focused when the Modal opens
+   */
+  selectorPrimaryFocus?: string;
+
+  /** Specify the CSS selectors that match the floating menus. */
+  selectorsFloatingMenus?: string[];
+
+  size?: 'xs' | 'sm' | 'md' | 'lg';
+}
+
+const ComposedModal = React.forwardRef<HTMLDivElement, ComposedModalProps>(
+  function ComposedModal(
+    {
+      ['aria-labelledby']: ariaLabelledBy,
+      ['aria-label']: ariaLabel,
+      children,
+      className: customClassName,
+      containerClassName,
+      danger,
+      isFullWidth,
+      onClose,
+      onKeyDown,
+      open,
+      preventCloseOnClickOutside,
+      selectorPrimaryFocus,
+      selectorsFloatingMenus,
+      size,
+      ...rest
+    },
+    ref
+  ) {
+    const prefix = usePrefix();
+    const [isOpen, setisOpen] = useState<boolean>(!!open);
+    const [prevOpen, setPrevOpen] = useState<boolean>(!!open);
+    const innerModal = useRef<HTMLDivElement>(null);
+    const button = useRef<HTMLButtonElement>(null);
+    const startSentinel = useRef<HTMLButtonElement>(null);
+    const endSentinel = useRef<HTMLButtonElement>(null);
+
+    if (open !== prevOpen) {
+      setisOpen(!!open);
+      setPrevOpen(!!open);
+    }
+
+    function handleKeyDown(evt: KeyboardEvent) {
+      if (evt.key === 'Escape') {
+        closeModal(evt);
+      }
+
+      onKeyDown(evt);
+    }
+
+    function handleMousedown(evt: MouseEvent) {
+      if (
+        !innerModal.current?.contains(evt.target as Node) &&
+        preventCloseOnClickOutside
+      ) {
+        return;
+      }
+      if (innerModal.current?.contains(evt.target as Node)) {
+        closeModal(evt);
+      }
+    }
+
+    function handleBlur({
+      target: oldActiveNode,
+      relatedTarget: currentActiveNode,
+    }) {
+      if (open && currentActiveNode && oldActiveNode && innerModal.current) {
+        const { current: bodyNode } = innerModal;
+        const { current: startSentinelNode } = startSentinel;
+        const { current: endSentinelNode } = endSentinel;
+        wrapFocus({
+          bodyNode,
+          startTrapNode: startSentinelNode,
+          endTrapNode: endSentinelNode,
+          currentActiveNode,
+          oldActiveNode,
+          selectorsFloatingMenus,
+        });
+      }
+    }
+
+    function closeModal(evt) {
+      if (!onClose || onClose(evt) !== false) {
+        setisOpen(false);
+      }
+    }
+
+    const modalClass = cx(
+      `${prefix}--modal`,
+      isOpen && 'is-visible',
+      customClassName,
+      danger && `${prefix}--modal--danger`
+    );
```

Comment:
we've been using this className standard to prevent undefined classes. Same comment for `containerClass` below
<img width="722" alt="image" src="https://user-images.githubusercontent.com/40550942/233187280-b6854272-456c-4ac2-9987-e22f4bdd7bc9.png">